src/main/java/carparks/parking/ParkingService.java

```diff
@@ -0,0 +1,190 @@
+		
+		ParkingEntry mostRecent = getMostRecentFinishedSession(plateNumber);
+		BigDecimal fee = mostRecent.getFee();
+		if (!currency.equals("PLN")) {
+			fee = getConvertedValue(fee, currency);
+		}
+		return new ParkingEntryDto(plateNumber, fee, SessionStatus.FINISHED);
+
+	}
+
+	public ParkingEntryDto getEntry(int id, String currency) {
+
+		ParkingEntry entry = parkingDao.findById(id);
+
+		if (entry != null) {
+			if (entry.getFinish() == null) {
+				BigDecimal feeSoFar = feeCalculator.calculateFee(entry);
+				
+				if (currency != "PLN") {
+					feeSoFar = getConvertedValue(feeSoFar, currency);
+				}
+				
+				return new ParkingEntryDto(entry.getPlateNumber(), feeSoFar, SessionStatus.IN_PROGRESS);
+			}
+			
+			BigDecimal fee = entry.getFee();
+			if (!currency.equals("PLN")) {
+				fee = getConvertedValue(fee, currency);
+			}
+			return new ParkingEntryDto(entry.getPlateNumber(), fee, SessionStatus.FINISHED);
+		}
+		
+		throw new IllegalStateException("This vehicle has never parked");
+	}
+	
+	public DailySummaryDto getDailySummary(String currency) {
+		
+		LocalDateTime start = LocalDate.now().atStartOfDay();
+		LocalDateTime end = LocalDate.now().atTime(LocalTime.MAX);
+		
+		List<ParkingEntry> regularEntries = parkingDao.findByFeeTypeAndFinishBetween(FeeType.REGULAR, start, end);
+		List<ParkingEntry> vipEntries = parkingDao.findByFeeTypeAndFinishBetween(FeeType.VIP, start, end);
+		
+		BigDecimal regularTurnover = feeCalculator.calculateDailyTurnover(regularEntries);
+		BigDecimal vipTurnover = feeCalculator.calculateDailyTurnover(vipEntries);
+		
+		if (!currency.equals("PLN")) {
+			regularTurnover = getConvertedValue(regularTurnover, currency);
+			vipTurnover = getConvertedValue(vipTurnover, currency);
+		}
+		
+		return new DailySummaryDto(regularTurnover, vipTurnover);
+		
+	}
+	
+	
+	public DailySummaryDto getDailySummary(LocalDate date, String currency) {
+		
+		if (date.isAfter(LocalDate.now())) {
+			throw new IllegalArgumentException("The given day is in the future");
+		}
+		
+		LocalDateTime start = date.atStartOfDay();
+		LocalDateTime end = date.atTime(LocalTime.MAX);
+		
+		List<ParkingEntry> regularEntries = parkingDao.findByFeeTypeAndFinishBetween(FeeType.REGULAR, start, end);
+		List<ParkingEntry> vipEntries = parkingDao.findByFeeTypeAndFinishBetween(FeeType.VIP, start, end);
+		
+		BigDecimal regularTurnover = feeCalculator.calculateDailyTurnover(regularEntries);
+		BigDecimal vipTurnover = feeCalculator.calculateDailyTurnover(vipEntries);
+		
+		if (!currency.equals("PLN")) {
+			regularTurnover = getConvertedValue(regularTurnover, currency);
+			vipTurnover = getConvertedValue(vipTurnover, currency);
+		}
+		
+		return new DailySummaryDto(regularTurnover, vipTurnover);
+		
+	}
+
+	private List<ParkingEntry> getStartedSessionsListByPlateNumber(String plateNumber) {
+
+		List<ParkingEntry> entries = parkingDao.findByPlateNumber(plateNumber);
+
+		return entries.stream().filter(e -> (e.getFinish() == null))
```

Comment:
Czy widzisz jakiś problem w tym podejściu? Co by się działo jakby ktoś parkował parę razy dziennie przez parę lat?

Reply:
Rozumiem, że chodzi o to, żeby filtrować dane raczej po stronie bazy danych niż pobierać potencjalnie dużą ilość wartości po to, żeby i tak zaraz większość z nich odrzucić. Optymalniejsze byłoby bardziej szczegółowe zapytanie do bazy danych.